src/VisualStudio/Core/Def/Implementation/IntellisenseControls/IntellisenseTextBoxViewModelFactory.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Projection;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Microsoft.VisualStudio.Utilities;
+using ImportingConstructorAttribute = System.Composition.ImportingConstructorAttribute;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.IntellisenseControls
+{
+    [Export(typeof(IntellisenseTextBoxViewModelFactory)), Shared]
+    internal class IntellisenseTextBoxViewModelFactory
+    {
+        private static readonly INITVIEW[] s_InitViews = new[] {
+            new INITVIEW()
+            {
+                fSelectionMargin = 0,
+                fWidgetMargin = 0,
+                fDragDropMove = 0,
+                IndentStyle = vsIndentStyle.vsIndentStyleNone
+            }
+        };
+
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly ITextEditorFactoryService _textEditorFactoryService;
+        private readonly OLE.Interop.IServiceProvider _serviceProvider;
+        private readonly IProjectionBufferFactoryService _projectionBufferFactoryService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public IntellisenseTextBoxViewModelFactory(
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            ITextEditorFactoryService textEditorFactoryService,
+            IContentTypeRegistryService contentTypeRegistryService,
+            IProjectionBufferFactoryService projectionBufferFactoryService,
+            SVsServiceProvider services)
+        {
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _textEditorFactoryService = textEditorFactoryService;
+            _serviceProvider = (OLE.Interop.IServiceProvider)services.GetService(typeof(OLE.Interop.IServiceProvider));
+            _projectionBufferFactoryService = projectionBufferFactoryService;
+        }
+
+        public async Task<IntellisenseTextBoxViewModel[]> CreateIntellisenseTextBoxViewModelsAsync(
+          Document document,
+          IContentType contentType,
+          string contentString,
+          Func<IProjectionSnapshot, ITrackingSpan[]> createSpansMethod,
```

Comment:
```suggestion
          Func<ITextSnapshot, ITrackingSpan[]> createSpansMethod,
```

It looks like the only reason this is taking an IProjectionSnapshot is because contextBuffer is one, but that's an implementation detail that doesn't seem applicable to callers?